### PR TITLE
feat(carto): Respect `matchingColumn` prop when merging boundary data

### LINKED ISSUE #8934: [Bug] CARTO VectorTileLayer doesn't respect `matchingColumn`
### Description

When using a `boundaryXXXSource` it is possible to specify a `matchingColumn` on which to perform the join. This is not working unless the `matchingColumn` is set to `'geoid'`. The cause is the hardcoded value in `mergeBoundaryData`:

https://github.com/visgl/deck.gl/blob/master/modules/carto/src/layers/utils.ts#L22

### Flavors

- [ ] Script tag
- [ ] React
- [ ] Python/Jupyter notebook
- [ ] MapboxOverlay
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] ArcGIS

### Expected Behavior

_No response_

### Steps to Reproduce

`matchingColumn` is respected

### Environment

- Framework version: deck.gl@9.0.14
- Browser: all
- OS: all

### Logs

_No response_

### PR DESCRIPTION
**Pull request by felixpalmer**
Closes #8934

#### Change List
- Pass `matchingColumn` prop down to `VectorTileLayer`
- Do not hardcode `'geoid'` in `mergeBoundaryData`
- Tests for `layer/utils`

### CODE REVIEW

File: modules/carto/src/layers/utils.ts [hunk 1]
```diff
@@ -16,13 +16,26 @@ export function injectAccessToken(loadOptions: any, accessToken: string): void {
...
+  properties: PropertiesTile,
+  matchingColumn: string
+): VectorTile {
   const mapping = {};
-  for (const {geoid, ...rest} of properties.properties) {
-    if (geoid in mapping) {
-      log.warn('Duplicate geoid key in boundary mapping, using first occurance')();
+  for (const property of properties.properties) {
+    const matchingKey = property[matchingColumn];
+    if (!matchingKey || matchingKey === '') {
```
Comment (donmccurdy) [inline modules/carto/src/layers/utils.ts:27]: nit, redundant falsy check:

```suggestion
    if (!matchingKey) {
```

File: modules/carto/src/layers/utils.ts [hunk 2]
```diff
@@ -16,13 +16,26 @@ export function injectAccessToken(loadOptions: any, accessToken: string): void {
...
+      log.warn(
+        `Properties: Missing value for matchingKey with matchingColumn: ${matchingColumn}`
+      )();
+    }
+    const row = {...property};
+    delete row[matchingColumn];
+
+    if (matchingKey in mapping) {
+      log.warn('Properties: Duplicate key in boundary mapping, using first occurance')();
```
Comment (donmccurdy) [inline modules/carto/src/layers/utils.ts:36]:
```suggestion
      log.warn('Properties: Duplicate key in boundary mapping, using first occurrence')();
```

File: modules/carto/src/layers/utils.ts [hunk 3]
```diff
@@ -32,7 +45,16 @@ export function mergeBoundaryData(geometry: VectorTile, properties: PropertiesTi
       continue;
     }
 
-    geom.properties = geom.properties.map(({geoid}) => mapping[geoid]);
+    geom.properties = geom.properties.map((property: Record<string, string>) => {
+      const matchingKey = property[matchingColumn];
+      if (!matchingKey || matchingKey === '') {
```
Comment (donmccurdy) [inline modules/carto/src/layers/utils.ts:50]:
```suggestion
      if (!matchingKey) {
```

File: modules/carto/src/layers/utils.ts [hunk 4]
```diff
@@ -16,13 +16,26 @@ export function injectAccessToken(loadOptions: any, accessToken: string): void {
...
-      log.warn('Duplicate geoid key in boundary mapping, using first occurance')();
+  for (const property of properties.properties) {
+    const matchingKey = property[matchingColumn];
+    if (!matchingKey || matchingKey === '') {
+      log.warn(
+        `Properties: Missing value for matchingKey with matchingColumn: ${matchingColumn}`
+      )();
+    }
+    const row = {...property};
+    delete row[matchingColumn];
```
Comment (donmccurdy) [inline modules/carto/src/layers/utils.ts:33]: What would happen if we didn't delete the matching column? Just wondering if we can avoid copying every row, and/or should leave a comment so we don't try to make that optimization incorrectly later.

File: modules/carto/src/sources/types.ts [hunk 5]
```diff
@@ -214,7 +214,7 @@ export interface VectorLayer {
   fields: Record<string, string>;
 }
 
-export type TilejsonResult = Tilejson & {accessToken: string};
+export type TilejsonResult = Tilejson & {accessToken: string; matchingColumn?: string};
```
Comment (donmccurdy) [inline modules/carto/src/sources/types.ts:217]: Is this the right place for `matchingColumn`? Could we instead have boundaryTableSource return `Promise<TilejsonResult & {matchingColumn: string}>`?

**Review by donmccurdy — APPROVED** [review]

File: modules/carto/src/layers/utils.ts (hunk 4 above)
Comment (felixpalmer) [reply modules/carto/src/layers/utils.ts:33]: Nothing much, we would just expose the `matchingColumn` to the layer accessors. I guess that doesn't hurt and might in fact be useful, e.g. visualizing the ZIP code

File: modules/carto/src/sources/types.ts (hunk 5 above)
Comment (felixpalmer) [reply modules/carto/src/sources/types.ts:217]: Yes, I started with that but it was getting a bit messy - but perhaps it is worth the extra typing though...

**Review by donmccurdy — APPROVED** [review]

**Comment by felixpalmer** [comment]
Closing in favour of https://github.com/visgl/deck.gl/pull/8971

**Pull request closed without merging**